darkcluster/src/main/java/com/linkedin/darkcluster/impl/DarkClusterStrategyFactoryImpl.java

```diff
@@ -0,0 +1,172 @@
+  private final Map<String, AtomicReference<DarkClusterStrategy>> _darkStrategyMap;
+  private final Random _random;
+  private final LoadBalancerClusterListener _clusterListener;
+  private final DarkClusterVerifierManager _verifierManager;
+
+  public DarkClusterStrategyFactoryImpl(@Nonnull Facilities facilities,
+                                        @Nonnull String sourceClusterName,
+                                        @Nonnull DarkClusterDispatcher darkClusterDispatcher,
+                                        @Nonnull Notifier notifier,
+                                        @Nonnull Random random,
+                                        @Nonnull DarkClusterVerifierManager verifierManager)
+  {
+    _facilities = facilities;
+    _sourceClusterName = sourceClusterName;
+    _notifier = notifier;
+    _darkStrategyMap = new ConcurrentHashMap<>();
+    _random = random;
+    _darkClusterDispatcher = darkClusterDispatcher;
+    _verifierManager = verifierManager;
+    _clusterListener = new DarkClusterListener();
+  }
+
+  @Override
+  public void start()
+  {
+    _facilities.getClusterInfoProvider().registerClusterListener(_clusterListener);
+  }
+
+  @Override
+  public void shutdown()
+  {
+    _facilities.getClusterInfoProvider().unregisterClusterListener(_clusterListener);
+  }
+
+
+
+  @Override
+  public DarkClusterStrategy getOrCreate(@Nonnull String darkClusterName, @Nonnull DarkClusterConfig darkClusterConfig)
+  {
+    if (!_darkStrategyMap.containsKey(darkClusterName))
+    {
+      _darkStrategyMap.putIfAbsent(darkClusterName, new AtomicReference<>(createStrategy(darkClusterName, darkClusterConfig)));
+    }
+    // it's theoretically possible for the Listener to remove the entry after the containsKey but before we retrieve it. Rather
+    // than adding synchronization between accessors of _darkStrategyMap, we will make each put or get resilient
+    return _darkStrategyMap.getOrDefault(darkClusterName, NO_OP_DARK_CLUSTER_STRATEGY).get();
+  }
+
+  /**
+   * In the future, additional strategies can be added, and the logic here can choose the appropriate one based on the config values.
+   */
+  private DarkClusterStrategy createStrategy(String darkClusterName, DarkClusterConfig darkClusterConfig)
+  {
+    if (darkClusterConfig.hasMultiplier() && darkClusterConfig.getMultiplier() > 0)
+    {
+      BaseDarkClusterDispatcher baseDarkClusterDispatcher = new BaseDarkClusterDispatcherImpl(darkClusterName, _darkClusterDispatcher,
+                                                                                              _notifier, _verifierManager);
+      return new RelativeTrafficMultiplierDarkClusterStrategy(_sourceClusterName, darkClusterName, darkClusterConfig.getMultiplier(),
+                                                              baseDarkClusterDispatcher, _notifier, _facilities.getClusterInfoProvider(), _random);
+    }
+    else
+    {
+      return new NoOpDarkClusterStrategy();
```

Comment:
actually, we do expect this to be returned in some cases. For example, if users have intentionally set their multiplier to zero because they don't want traffic to go to the dark cluster.  This can be a temporary thing, so it is an expected use case. I added a comment.